tree/ntuple/v7/inc/ROOT/RNTupleJoinTable.hxx

```diff
@@ -35,63 +36,114 @@ namespace Internal {
+
+      /// The size (in bytes) for each join field, corresponding to `fJoinFieldNames`. This information is stored to be
+      /// able to properly cast incoming void pointers to the join field values in `GetEntryIndexes`.
+      std::vector<std::size_t> fJoinFieldValueSizes;
+
+   public:
+      //////////////////////////////////////////////////////////////////////////
+      /// \brief Build the entry mapping.
+      void Build();
+
+      //////////////////////////////////////////////////////////////////////////
+      /// \brief Get the entry indexes for this entry mapping.
+      const std::vector<ROOT::NTupleSize_t> *GetEntryIndexes(std::vector<void *> valuePtrs) const;
+
+      //////////////////////////////////////////////////////////////////////////
+      /// \brief Create a new entry mapping.
+      ///
+      /// \param[in] pageSource The page source of the RNTuple with the entries to map.
+      /// \param[in] joinFieldNames Names of the join fields to use in the mapping.
+      REntryMapping(RPageSource &pageSource, const std::vector<std::string> &joinFieldNames)
+         : fPageSource(&pageSource), fJoinFieldNames(joinFieldNames)
+      {
+      }
+   };
    /// Names of the join fields used for the mapping to their respective entry indexes.
    std::vector<std::string> fJoinFieldNames;
 
-   /// The size (in bytes) for each join field, corresponding to `fJoinFieldNames`. This information is stored to be
-   /// able to properly cast incoming void pointers to the join field values in `GetEntryIndexes`.
-   std::vector<std::size_t> fJoinFieldValueSizes;
+   /// Partition keys used in the join table.
+   std::vector<PartitionKey_t> fPartitionKeys;
+
+   /// Partitions of one or multiple entry mappings.
+   std::vector<std::vector<std::unique_ptr<REntryMapping>>> fPartitions;
 
    /// Only built join tables can be queried.
    bool fIsBuilt = false;
 
    /////////////////////////////////////////////////////////////////////////////
    /// \brief Create an a new RNTupleJoinTable for the RNTuple represented by the provided page source.
    ///
-   /// \param[in] fieldNames The names of the join fields to use for the join table. Only integral-type fields are
+   /// \param[in] joinFieldNames The names of the join fields to use for the join table. Only integral-type fields are
    /// allowed.
-   RNTupleJoinTable(const std::vector<std::string> &fieldNames) : fJoinFieldNames(fieldNames) {}
+   RNTupleJoinTable(const std::vector<std::string> &joinFieldNames) : fJoinFieldNames(joinFieldNames) {}
 
    /////////////////////////////////////////////////////////////////////////////
-   /// \brief Ensure the RNTupleJoinTable has been built.
-   ///
-   /// \throws RException If the join table has not been built, and can therefore not be used yet.
-   void EnsureBuilt() const;
+   /// \brief Get all entry mappings in the join table.
+   const std::vector<REntryMapping *> GetAllMappings() const;
```

Comment:
Since you're returning a value, it's better to return it non-const (same thing for the next function)